### PR TITLE
Make sure the audit log is written if the DB is updated.

### PR DESCRIPTION
**Pull request by fabricematrat**

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1048/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1049/
Test PASSed.

File: internal/router/fieldinclude.go [hunk 1]
```diff
@@ -17,14 +18,18 @@ type FieldQueryFunc func(id *ResolvedURL, selector map[string]int, req *http.Req
 
 // FieldUpdater records field changes made by a FieldUpdateFunc.
 type FieldUpdater struct {
-	fields map[string]interface{}
-	search bool
+	fields  map[string]interface{}
+	entries []interface{}
```
Comment (frankban) [inline internal/router/fieldinclude.go]: Could this be a []*audit.Entry?

File: internal/router/fieldinclude.go [hunk 2]
```diff
@@ -35,7 +40,7 @@ func (u *FieldUpdater) UpdateSearch() {
 // A FieldUpdateFunc is used to update a metadata document for the
 // given id. For each field in fields, it should set that field to
 // its corresponding value in the metadata document.
-type FieldUpdateFunc func(id *ResolvedURL, fields map[string]interface{}) error
+type FieldUpdateFunc func(id *ResolvedURL, fields map[string]interface{}, entries []interface{}) error
```
Comment (frankban) [inline internal/router/fieldinclude.go]: ...and this too?

**Comment by frankban** [comment]
:+1: with a question. Thank you!

File: internal/v4/api.go [hunk 3]
```diff
@@ -308,18 +308,40 @@ func (h *ReqHandler) puttableBaseEntityHandler(get baseEntityHandlerFunc, handle
 	})
 }
 
-func (h *ReqHandler) updateBaseEntity(id *router.ResolvedURL, fields map[string]interface{}) error {
+var testAddAuditCallback func(e audit.Entry)
+
+func (h *ReqHandler) processEntries(entries []interface{}) {
+	for _, v := range entries {
+		if e, err := v.(audit.Entry); err {
```
Comment (mhilton) [inline internal/v4/api.go]: s/err/ok/g

File: internal/v4/api.go [hunk 4]
```diff
@@ -308,18 +308,40 @@ func (h *ReqHandler) puttableBaseEntityHandler(get baseEntityHandlerFunc, handle
 	})
 }
 
-func (h *ReqHandler) updateBaseEntity(id *router.ResolvedURL, fields map[string]interface{}) error {
+var testAddAuditCallback func(e audit.Entry)
+
+func (h *ReqHandler) processEntries(entries []interface{}) {
```
Comment (mhilton) [inline internal/v4/api.go]: Are these realistically ever going to be anything other than audit log entries? probably better to use the type system here and not do unnecessary type assertions.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1050/
Test PASSed.

**Comment by mhilton** [comment]
:+1: LGTM

**Comment by fabricematrat** [comment]
:shipit: thanks @mhilton && @frankban

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Pull request merged**